### PR TITLE
Fix event validation when adding and updating events

### PR DESCRIPTION
**Pull request by tej99**
# Motivation and Context

Currently the validation is not flexible enough to allow validation
on both update and creation. This PR makes the validation work with
both create events and update events. Also making sure all but two
 business rules (24hrs apart and dates are in the future) are captured 
within the validation.

# What has changed

Add 'complete' event validation when both updating and creating events

# How to test?

Ran unit tests, ITs, ATs

Can use this response operation branch if you prefer clicking around over api calls>> https://github.com/ONSdigital/response-operations-ui/pull/223

Double check business rules covered:
* The user can only input Main Print Selection date (MPS) that ~~is at least 24 hours~~ before the Go Live Date 
* The user can only input a Go Live Date that is after Main Print Selection but ~~at least 24 hours~~ before the Return By Date
* The user can only enter a Return By Date that is after the Go Live Date but ~~at least 24 hours~~ before the Collection Exercise End Date
* The user can only enter a Collection Exercise End Date that is after the Return By Date
* ~~All Mandatory dates must also be in the future~~

* ~~All Optional event dates can not be set in the past apart from Reference Start and End dates & Employment dates~~
* The user can only enter a Reminder Date that is ~~at least 24 hours~~ after the Go Live Date and ~~at least 24 hours~~ before the Collection Exercise End Date
* Second reminder date must be ~~at least 24 hours~~after the first reminder date and ~~at least 24 hours~~ before Collection end date
* Third reminder date must be ~~at least 24 hours~~ after the second reminder date and ~~at least 24 hours~~ before Collection end date
* The reference period end date must always be after the reference period start date
* Employment date has no validation

* The mandatory business event dates are editable but they cannot be changed once the state of a collection exercise is Ready for Live (and all other states that indicate the CE is/has been executed)

# Links

### CODE REVIEW

**Comment by tej99** [comment]
[Add event validation on collection exercise execution](https://trello.com/c/ovBAppnQ/172-add-event-validation-on-collection-exercise-execution)

**Comment by tej99** [comment]
[Add events validation for creating and updating events](https://trello.com/c/Ig1gQMls/171-add-events-validation-for-creating-and-updating-events)

File: src/test/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpointIT.java [hunk 1]
```diff
@@ -647,27 +646,22 @@ private UUID createScheduledCollectionExercise() throws CTPException, IOExceptio
...
-    for (EventService.Tag t : tags) {
-      EventDTO event = new EventDTO();
-      event.setCollectionExerciseId(collectionExerciseId);
-      event.setTag(t.name());
-      // mandatory dates must be minimum of 24 hours apart
-      event.setTimestamp(Timestamp.from(Instant.now().plus(days, ChronoUnit.DAYS)));
-      this.client.createCollectionExerciseEvent(event);
-      days += 2;
-    }
+    Arrays.stream(EventService.Tag.values())
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpointIT.java]: 👍

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/BusinessEventValidator.java [hunk 2]
```diff
@@ -19,74 +15,99 @@
...
+
+    if (!eventDuringExercise(goLive, newEvent, exerciseEnd)) {
       return false;
     }
 
-    Map<String, Event> eventMap = generateEventsMap(existingEvents, updatedEvent);
+    final Event currentReminder = eventMap.get(newEvent.getTag());
 
-    return validateMandatoryEvents(eventMap) && validateNonMandatoryEvents(eventMap);
+    if (currentReminder != null && isEventInPast(currentReminder)) {
```
[the remark this answers is not in the document]
Comment (tej99) [reply src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/BusinessEventValidator.java]: eventMap (existing events) may not have the newEvent (submitted event) tag if that event hasn't been created yet. I should probably go back and rename some of these variables 🤔

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/BusinessEventValidator.java [hunk 3]
```diff
@@ -19,74 +15,99 @@
...
-    // Can only update reminders of the non mandatory events when READY_FOR_LIVE
-    if ((collectionExerciseState.equals(CollectionExerciseState.READY_FOR_LIVE)
-            || collectionExerciseState.equals(CollectionExerciseState.LIVE))
-        && (isMandatory(updatedEvent) || !isReminder(updatedEvent))) {
-      return false;
+    final Map<String, Event> events =
+        existingEvents.stream().collect(Collectors.toMap(Event::getTag, Function.identity()));
+
+    if (isMandatory(submittedEvent)) {
+      return validateMandatoryEvents(events, submittedEvent, collectionExerciseState);
```
Comment (PurpleBooth) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/BusinessEventValidator.java]: These looks like "types" of validators that belong in their own classes

Comment (tej99) [reply src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/BusinessEventValidator.java]: Done

**Pull request merged**